stable/aws-for-fluent-bit/templates/service.yaml

```diff
@@ -11,6 +11,14 @@ spec:
     port: {{ .Values.serviceMonitor.service.port }}
     protocol: TCP
     targetPort: {{ .Values.serviceMonitor.service.targetPort }}
+{{- if .Values.serviceMonitor.service.extraPorts }}
```

Comment:
I think the template for the extraEndpoints below needs to be similar to this, explicitly model each field.

Reply:
@PettitWesley  the template is correct I double checked now. I have made the ammendmends on the instrustions on how to do the the test run (Should have included that in PR message too):

## Changes in Values
* Before testing the chart, the `values.yaml` file needs to be updated as below: 
> NB: (Note that we remove the `[]` on lines 387 and 371 and set the service monitor to enabled
```
serviceMonitor:
  service: 
    type: ClusterIP
    port: 2020
    targetPort: 2020
    extraPorts: # Removing the empty array placeholder []  because we are now pupulating the array
      - port: 2021
        targetPort: 2021
        protocol: TCP
        name: metrics
  ## When set true then use a ServiceMonitor to configure scraping
  enabled: true # Enabling the servicMonitor
  interval: 30s
  telemetryPath: /api/v1/metrics/prometheus
  labels:
    # app: example-application
    # teamname: example
  timeout: 10s
  relabelings: []
  targetLabels: []
  metricRelabelings: []
  extraEndpoints:  # removing the empty array placeholder [] because now we are populating the array
    - port: metrics
      path: /metrics
      interval: 30s
      scrapeTimeout: 10s
      scheme: http
```
* In addition, the `serviceMonitor` object is a CRD created by the promehteus community. If your are doing helm template locally using `helm 3.x` please see below:

## Templating locally
Helm does not automatically recognise CRDs, to be able to output the template including CRDs we need to:
* Download the `serviceMonitor` CRD from the prometheus operator Repo: https://github.com/prometheus-operator/prometheus-operator/blob/main/example/prometheus-operator-crd-full/monitoring.coreos.com_servicemonitors.yaml
* We add this YAML file to our chart directory either in the root together with Chart.yamland values.yaml or under the templates directory. This will help helm to understand the CRD resource.
* Next in the `serviceMonitor.yaml` template on line two, there is a condition: `{{- if and ( .Capabilities.APIVersions.Has "monitoring.coreos.com/v1" ) ( .Values.serviceMonitor.enabled ) }}` -  which checks if the cluster API version has the "monitoring.coreos.com/v1" capabillity. For a dry run, since we are not installing this on an actual cluster with the relevant API version installed, we may temporarily disable that check by changing the `and` to an `or`. (If you are test installing on actual cluster -  please see the next heading.)
* Once all this is setup -  you can then run `helm template .` and inspect the output for `kind: Service` and the `kind: ServiceMonitor`
* Revert the condition in `serviceMonitor.yaml` template back to `and` AND  delete the serviceMonitor CRD file from the chart directory
* In values.yaml -  comment out lines [372-375] amnd [388 - 392], place back the the empty array place holders `[]` on lines 371 and 387, and set the `enabled: false` for the serviceMonitor such that theconfig looks like below:

```
serviceMonitor:
  service: 
    type: ClusterIP
    port: 2020
    targetPort: 2020
    extraPorts: []
      # - port: 2021
      #   targetPort: 2021
      #   protocol: TCP
      #   name: metrics
  ## When set true then use a ServiceMonitor to configure scraping
  enabled: false
  interval: 30s
  telemetryPath: /api/v1/metrics/prometheus
  labels:
    # app: example-application
    # teamname: example
  timeout: 10s
  relabelings: []
  targetLabels: []
  metricRelabelings: []
  extraEndpoints: []
    # - port: metrics
    #   path: /metrics
    #   interval: 30s
    #   scrapeTimeout: 10s
    #   scheme: http
``` 

## Installing on test cluster
If you are testing by installing the helm chart on a cluster -  you need to ensure that you first install the `prometheus-operator` CRDs on the cluster -  especially the `serviceMonitor` in our case.

### Example using minikube cluster
* Download the serviceMonitor CRD from the prometheus operator repo: https://github.com/prometheus-operator/prometheus-operator/blob/main/example/prometheus-operator-crd-full/monitoring.coreos.com_servicemonitors.yaml
* Run `kubectl apply -f </path/to/serviceMonitor.yaml>` to install the CRD on the cluster (this is so that the condition in serviceMonitor.yaml template passes)
* In the values.yaml of our chart:
    * Enable the serviceMonitor
    * Uncomment the lines [372-375] and lines [388-392]
    * remove the empty array placeholders `[]` on lines 387 and 371
* The config for service Monitor should now look like:

```
serviceMonitor:
  service: 
    type: ClusterIP
    port: 2020
    targetPort: 2020
    extraPorts:
      - port: 2021
        targetPort: 2021
        protocol: TCP
        name: metrics
  ## When set true then use a ServiceMonitor to configure scraping
  enabled: true
  interval: 30s
  telemetryPath: /api/v1/metrics/prometheus
  labels:
    # app: example-application
    # teamname: example
  timeout: 10s
  relabelings: []
  targetLabels: []
  metricRelabelings: []
  extraEndpoints: 
    - port: metrics
      path: /metrics
      interval: 30s
      scrapeTimeout: 10s
      scheme: http
```

* Run `helm install aws-for-fluentbit .` from the chart directory. Helm will install the chart.
* To get the service monitor run `kubectl get serviceMonitor -n <namespace where chart is deployed>` and you will see the service monitor.
* Run `kubectl describe serviceMonitor <serviceMonitor name from previous output> -n <the namespace where it is installed>` and inspect the out put.
* To see the yaml output run: `kubectl get serviceMonitor <serviceMonitor name> -n <namespace of serviceMonitor> -o yaml` and inspect the output
* Do the same for the `Service` resource